File: Challenges/1-4th_grade_math_problem/molo/challenge.py
import sys
import itertools

for p in itertools.permutations([1, 2, 3, 4, 5, 6, 7, 8, 9]):
    """
    In order to cut down on the numbers to check (almost a billion numbers),
    Only iterate through the permutations, where there are only 9!

    Then, as we loop through each number, check its divisibility.
    We iteratively make the number larger. For example, we first
    check divisibility of the first number, then the second number
    we can take the previous, multiply by ten and add the next.

    If we never break out of the for loop, we move to the else
    statement and announce that we found the answer.
    """

    check = 0
    for idx, val in enumerate(p):
        check = (check * 10) + val

        if check % (idx + 1) != 0:
            break
    else:
        print("Found! %d" % check)
        sys.exit()

File: Challenges/C++ Challenges/1-4th_grade_math_problem/molo/challenge.c
#include <stdio.h>
#include <stdlib.h>

void permutations(int* options, int size, int original_size) {
    if (size == 1) {

        int check = options[0];
        for (int i = 1; i < original_size; i++) {
            check = check*10 + options[i];

            if (check % (i+1) != 0) {
                return;
            }
        }

        printf("Found! %d\n", check);
        exit(0);
    }

    for (int i = 0; i < size; i++) {
        permutations(options, size-1, original_size);

        int j = size % 2 ? i : 0;

        int tmp = options[size-1];
        options[size-1] = options[j];
        options[j] = tmp;
    }
}

int main() {
    int options[] = {1, 2, 3, 4, 5, 6, 7, 8, 9};
    const int size = sizeof(options)/sizeof(options[0]);

    permutations(options, size, size);
}
